flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/InternalTimerService.java

```diff
@@ -18,43 +18,306 @@
+	}
 
 	/** Returns the current processing time. */
-	long currentProcessingTime();
+	public long currentProcessingTime() {
+		return processingTimeService.getCurrentProcessingTime();
+	}
 
 	/** Returns the current event-time watermark. */
-	long currentWatermark();
+	public long currentWatermark() {
+		return currentWatermark;
+	}
 
 	/**
 	 * Registers a timer to be fired when processing time passes the given time. The namespace
 	 * you pass here will be provided when the timer fires.
 	 */
-	void registerProcessingTimeTimer(N namespace, long time);
+	abstract public void registerProcessingTimeTimer(N namespace, long time);
 
 	/**
 	 * Deletes the timer for the given key and namespace.
 	 */
-	void deleteProcessingTimeTimer(N namespace, long time);
+	abstract public void deleteProcessingTimeTimer(N namespace, long time);
 
 	/**
 	 * Registers a timer to be fired when processing time passes the given time. The namespace
 	 * you pass here will be provided when the timer fires.
 	 */
-	void registerEventTimeTimer(N namespace, long time);
+	abstract public void registerEventTimeTimer(N namespace, long time);
 
 	/**
 	 * Deletes the timer for the given key and namespace.
 	 */
-	void deleteEventTimeTimer(N namespace, long time);
+	abstract public void deleteEventTimeTimer(N namespace, long time);
+
+	/**
+	 * Returns the timers for the given key group.
+	 */
+	abstract public Set<InternalTimer<K, N>> getEventTimeTimersForKeyGroup(int keyGroup);
+
+	/**
+	 * Returns the timers for the given key group.
+	 */
+	abstract public Set<InternalTimer<K, N>> getProcessingTimeTimersForKeyGroup(int keyGroup);
+
+	/**
+	 * Restores the timers for the given key group.
+	 */
+	abstract public void restoreEventTimeTimersForKeyGroup(int keyGroup, Iterable<InternalTimer<K, N>> timers);
+
+	/**
+	 * Restores the timers for the given key group.
+	 */
+	abstract public void restoreProcessingTimeTimersForKeyGroup(int keyGroup, Iterable<InternalTimer<K, N>> timers);
+
+	/**
+	 * Starts the execution of the timer service
+	 */
+	abstract public void start();
+
+	/**
+	 * Closes the timer service.
+	 */
+	abstract public void close();
+	
+	public void advanceWatermark(long watermark) throws Exception {
+		if (watermark < currentWatermark) {
+			throw new IllegalStateException("The watermark is late.");
+		}
+		
+		currentWatermark = watermark;
+		
+		onEventTime(watermark);
+	}
+
+	/**
+	 * Snapshots the timers (both processing and event time ones) for a given {@code keyGroupIdx}.
+	 * @param stream the stream to write to.
+	 * @param keyGroupIdx the id of the key-group to be put in the snapshot.
+	 */
+	public void snapshotTimersForKeyGroup(DataOutputViewStreamWrapper stream, int keyGroupIdx) throws Exception {
+		InstantiationUtil.serializeObject(stream, keySerializer);
+		InstantiationUtil.serializeObject(stream, namespaceSerializer);
+
+		// write the event time timers
+		Collection<InternalTimer<K, N>> eventTimers = getEventTimeTimersForKeyGroup(keyGroupIdx);
```

Comment:
We could refactor out a method that writes `Collection<InternalTimer<K, N>>` and call it twice to de-duplicate code.